### PR TITLE
Disable unused/insecure php functions

### PR DESCRIPTION
**Pull request by Timdebruijn**
Disable unused/insecure php functions

### CODE REVIEW

**Comment by tvdijen** [comment]
There is no way of telling what functions are unused in EB, because of AM...
Also, how insecure they are is dependent on implementation/context.. They can be perfectly safe within a sandboxed environment.. If these functions can be abused, security has already failed 10 steps earlier

**Comment by Timdebruijn** [comment]
You would only need a breach in the php code and upload some sort of malicious code, than you would be able to use functions like exec, proc_open and curl_multi_exec.

On the other hand you are telling; well if they already broke a few layers of security, we should not do anything about that? Lets not stop them from using exec, proc_open...
Whats wrong with extra layers of security?

**Comment by tvdijen** [comment]
> You would only need a breach in the php code and upload some sort of malicious code, than you would be able to use functions like exec, proc_open and curl_multi_exec.
If that's what you're worried about, then maybe setting `allow_url_fopen = Off` makes more sense and is more of a generic solution.
 
> Whats wrong with extra layers of security?
Nothing, it's just that I think the suggested change is sub-optimal as the list of functions may vary depending on PHP-version, loaded extensions, etc.. And then keeping the list up-to-date becomes a security-risk itself.

**Comment by quartje** [comment]
I would say that it is good security practice to disable any unwanted functions. If you want to override these suggested disabled functions you can. 
I'll investigate whether we can set allow_url_fopen = Off as well, seems like a good suggestion.

**Comment by tvdijen** [comment]
> I'll investigate whether we can set allow_url_fopen = Off as well, seems like a good suggestion.

Composer requires it to be on, so for the fpm-pool file it's fine to disable, but if you want to disable it globally it would require a change to every call to composer (adding `-d allow_url_fopen=1` to the command).

File: roles/engineblock/templates/engine-pool.conf.j2 [hunk 1]
```diff
@@ -223,3 +223,4 @@ php_admin_flag[log_errors] = on
 ; Set session path to a directory owned by process user
 php_value[session.save_handler] = files
 php_value[session.save_path] = {{ php_session_dir}}/engine
+php_value[disable_functions] = exec,passthru,shell_exec,system,proc_open,popen,curl_multi_exec,show_source,pcntl_alarm,pcntl_fork,pcntl_waitpid,pcntl_wait,pcntl_wifexited,pcntl_wifstopped,pcntl_wifsignaled,pcntl_wifcontinued,pcntl_wexitstatus,pcntl_wtermsig,pcntl_wstopsig,pcntl_signal,pcntl_signal_dispatch,pcntl_get_last_error,pcntl_strerror,pcntl_sigprocmask,pcntl_sigwaitinfo,pcntl_sigtimedwait,pcntl_exec,pcntl_getpriority,pcntl_setpriority,
```
[the remark this answers is not in the document]
Comment (Timdebruijn) [reply roles/engineblock/templates/engine-pool.conf.j2]: Yes is does! When you request a page with phpinfo() you can see that the value's are added correctly.

**Comment by tvdijen** [comment]
We may also consider setting `session.use_strict_mode=1` globally

**Comment by tvdijen** [comment]
Note that `proc_open()` and `pcntl_signal()` are being used by composer

**Comment by thijskh** [comment]
I think this is useful. The mentioned functions should not be used from AM's. I would limit this to the FPM only, and leave cli as-is. There's no real benefit there I think, and indeed complicates things like running composer.

**Comment by quartje** [comment]
closing this PR, I've opened another one to continue this work (https://github.com/OpenConext/OpenConext-deploy/pull/334)

**Pull request closed without merging**